torch/csrc/autograd/engine.cpp

```diff
@@ -729,15 +649,30 @@ auto Engine::execute(const edge_list& roots,
   ClearCallbacks _cb_guard(final_callbacks_, post_callbacks_lock_);
 
   GraphTask graph_task(keep_graph, create_graph, worker_device == NO_DEVICE ? 0 : total_depth+1);
-  // Lock mutex while GraphTask is being set up
-  std::unique_lock<std::mutex> lock(graph_task.mutex_);
 
   // Now compute the dependencies for all executable functions and queue the root
   auto graph_root = std::make_shared<GraphRoot>(roots, inputs);
   compute_dependencies(graph_root.get(), graph_task);
+
   if (!outputs.empty()) {
     graph_task.init_to_execute(*graph_root, outputs);
   }
+  return execute_with_graph_task(graph_task, graph_root);
+}
+
+void Engine::enqueue_on_cpu(NodeTask task) {
+  std::call_once(start_threads_flag_, &Engine::start_threads, this);
```

Comment:
I'm not sure why we're spraying a `start_threads` initialization here. The invariant for this function is that, there must be a autograd job already running, right? We should just assert on that instead.

Reply:
OK